arduinolux/src/Debug/DebugTest.cpp

```diff
@@ -0,0 +1,14 @@
+#include"../StringConverter.h"
+#include <string>
+#include <iostream>
+
+int main() {
+	std::string testString = "Hello there world!";
+	do {
+		char* convertedString = StringConverter::ExtractCharArrayFromString(testString);
+
+		std::cout << convertedString << "\n";
```

Comment:
Ah. So, this is a "pass by value" vs "pass by reference" issue. If you make your function signature "inline char *foo(std::string arg)", and then you call it, like:

std::string mystring = "some data";
char *result = foo(mystring);

then what happens is, a new object is created from mystring at the time you call foo, and that new object is what 'arg' refers to in the body of the function. But, when the function returns, that object goes out of scope and is destroyed. So by the time you assign the result of arg.c_str() to result, it is referencing invalid memory (because that's how std::string::c_str() works, it gives you a pointer into the memory owned by the std::string object; object destroyed, memory no longer valid.)

If instead your function signature is 'inline char *foo(std::string &arg)' that means "pass a reference to the object I pass in as an argument down into the body of the function" and then it should work.

Of course, if you do this:

char *result;
{
  std::string mystring = "some text";
  result = foo(mystring);
}
cout << result << "\n";

then you're back to it not working again, because by the time you try to *use* result, mystring has gone out of scope and been destroyed.

Does that make sense? I highly recommend "Effective C++" by Scott Meyers to help get a handle on some of these sorts of issues.